src/utils/clone-with.ts

```diff
@@ -1,6 +1,6 @@
 import { TJSONSchema } from '../types/json-schema';
 import { clone } from './clone';
 
-export function cloneWith<T extends TJSONSchema = TJSONSchema> (schema: T, overrides: Partial<T>): T {
+export function cloneWith<T extends TJSONSchema = TJSONSchema, K extends TJSONSchema = TJSONSchema> (schema: T, overrides: K): T & K {
```

Comment:
I think this is the more correct type, overrides don't need to actually be `Partial<T>`, but instead could be anything, that could possibly overlap with T, but doesn't have to. The result is T & K, which lodash does agree on with me 😅

Reply:
`Partial<T>` tells the compiler that the overrides must correspond to the same type as the original schema. You should not be able to add `items: integer()` to a string schema for example (through it will work since schemas are allowed to contain any key/value as per the spec). I do agree that types could be better though, see below.

```typescript
declare function cloneWithOriginal<T extends TJSONSchema>(schema: T, overrides: Partial<T>): T;

const str = string(); // TStringJSONSchema

const clonedOriginal = cloneWithOriginal(str, { minLength: 3 }); // minLength still appears as optional in the cloned type: wrong

declare function cloneWithPullRequest<T extends TJSONSchema, O extends TJSONSchema = TJSONSchema>(schema: T, overrides: O): T & O;

const clonedPullRequest = cloneWithPullRequest(str, { minLength: 3 }); // Here you allow types like TStringJSONScheme & TArrayJSONSchema: wrong

declare function cloneWithBetter<T extends TJSONSchema, O extends Partial<T> = T>(schema: T, overrides: O): T & O;

const clonedBetter = cloneWithBetter(str, { minLength: 3 }); // Here we don't mix types and minLength has become non optional: better
```

Reply:
You also get autocompletion for the "better" types ;)

Reply:
`O extends Partial<T>` makes more sense. I got confused by the original test case that didn't have attributes in common 🙌 

```
it('should return a cloned object schema', () => {
    const schema = Schema.object({
      foo: Schema.string()
    }, {
      required: ['foo'],
      additionalProperties: true
    });

    const cloned = Schema.cloneWith(schema, Schema.object({ bar: Schema.string() }, {
      additionalProperties: true,
      nullable: true
    }));

    expect(cloned).to.deep.equal({
      type: ['object', 'null'],
      required: ['foo'],
      additionalProperties: true,
      properties: {
        bar: { type: 'string' }
      }
    });
  });
```